hassil/recognize.py

```diff
@@ -314,174 +341,175 @@ def _remove_skip_words(text: str, skip_words: Iterable[str]) -> str:
+            context_text = context_text[len(chunk_text) :]
             context_text = context_text.lstrip()
             yield MatchContext(
                 text=context_text,
                 # Copy over
-                slot_lists=context.slot_lists,
-                expansion_rules=context.expansion_rules,
                 entities=context.entities,
                 intent_context=context.intent_context,
             )
-        elif (match := NON_WORD_START.match(context.text)) is not None:
-            # Remove non-word characters and try again
-            context_text = context.text[len(match[0]) :].lstrip()
+        else:
+            # Remove punctuation and try again
+            context_text = PUNCTUATION.sub(" ", context.text).lstrip()
             if context_text.startswith(chunk_text):
                 context_text = context_text[len(chunk_text) :]
                 context_text = context_text.lstrip()
                 yield MatchContext(
                     text=context_text,
                     # Copy over
-                    slot_lists=context.slot_lists,
-                    expansion_rules=context.expansion_rules,
                     entities=context.entities,
                     intent_context=context.intent_context,
                 )
-
     elif isinstance(expression, Sequence):
         seq: Sequence = expression
         if seq.type == SequenceType.ALTERNATIVE:
             # Any may match (words | in | alternative)
             # NOTE: [optional] = (optional | )
             for item in seq.items:
-                yield from match_expression(context, item)
+                yield from match_expression(settings, context, item)
+
+        elif seq.type == SequenceType.GROUP:
+            if seq.items:
+                # All must match (words in group)
+                group_contexts = [context]
+                for item in seq.items:
+                    # Next step
+                    group_contexts = [
+                        item_context
+                        for group_context in group_contexts
+                        for item_context in match_expression(
+                            settings, group_context, item
+                        )
+                    ]
+                    if not group_contexts:
+                        break
 
-        elif seq.type == SequenceType.GROUP and seq.items:
-            # All must match (words in group)
-            group_contexts = [context]
-            for item in seq.items:
-                # Next step
-                group_contexts = [
-                    item_context
-                    for group_context in group_contexts
-                    for item_context in match_expression(group_context, item)
-                ]
-                if not group_contexts:
-                    break
-
-            for group_context in group_contexts:
-                yield group_context
+                for group_context in group_contexts:
```

Comment:
Why do list comprehension before yielding instead of yielding inside a for loop?

Reply:
A group should only yield a match once every item has individually matched in series. `group_contexts` will be empty if any part of the series fails to match.